### ISSUE TITLE
ThreadSanitizer catching race condition between "usrsctp_connect" and "usrsctp_bind"

### ISSUE THREAD

**Issue by dkisabaka** [sctplab/usrsctp#359]
This was caught by ThreadSanitizer while working with the webrtc DataChannel. In practice this happens very rarely rarely but it's easy to trigger with ThreadSanitizer

The fix is to guard `stcb->sctp_ep` when setting the connecting flag: https://github.com/dkisabaka/usrsctp/blob/ffa4ce549f328e29c395fd1913bd98cad020c15e/usrsctplib/netinet/sctp_usrreq.c#L8174

**Comment by tuexen** [comment]
Question: Why are you running `usrsctp_bind()` and `usrsctp_connect()` in two threads on the same end-point?

I'm not arguing if the above is a bug or not, I just want to understand where this can happen in practice...

**Comment by dkisabaka** [comment]
Two clients sending and receiving data through the data channel (on the same process), each client is doing so on a different thread. This can happen in unit tests for example: https://bugs.chromium.org/p/webrtc/issues/detail?id=10268

**Comment by tuexen** [comment]
I don't see how it makes sense. Your application is up to an race. `usrsctp_connect()` does in implicit bind if the socket is unbound. In this case your `usrsctp_bind()` call will fail. In my view, if you are wanting to do a `bind()` call and a `connect()` call, the `bind()` call must complete before you call `connect()`. In other cases the code sometimes does not what you expect...

**Comment by dkisabaka** [comment]
Right, thanks for the explanation. The issue seems to lie in how we're using those two methods. I'll update the WebRTC usage instead.

**Comment by taylor-b** [comment]
@tuexen I'm not sure if this was clear, but `usrsctp_connect()` and `usrsctp_bind()` are being called on two separate sockets.

One thread is here, in `sctp_endpoint_probe`:
```
LIST_FOREACH(inp, head, sctp_hash) {
	SCTP_INP_RLOCK(inp);
	if (inp->sctp_flags & SCTP_PCB_FLAGS_SOCKET_ALLGONE) {
```

While the other is here, in `sctpconn_connect`:
```
if (stcb->sctp_ep->sctp_flags & SCTP_PCB_FLAGS_TCPTYPE) {
	stcb->sctp_ep->sctp_flags |= SCTP_PCB_FLAGS_CONNECTED;
	/* Set the connected flag so we can queue data */
	soisconnecting(so);
}
```

The former holds the RLOCK to read `sctp_flags`, but the latter doesn't hold any lock to write it. Should it? Or is this type of race acceptable and should be ignored?

@dkisabaka, can you reopen this?

**Comment by tuexen** [comment]
I guess in principle we should own the write lock, but the connect() routines don't even hold the read lock. I guess we can clean this up. Maybe rrs@ remembers why he was not acquiring an inp lock.

**Comment by taylor-b** [comment]
I also notice that `sctp_ep->sctp_flags` is never read with a read lock. Is this something where you don't really care about races? Should we just suppress the warning and move on?

**Comment by apolyudov** [comment]
Is it feasible to make sctp_flags atomic (and use atomic bit operations on it) ? this should fix the data race.

**Comment by taylor-b** [comment]
I started looking into this. It looks like the INP locks are very rarely and briefly held. `sctp_findassociation_ep_addr` returns with the STCB lock held but not the INP lock. I imagine this is by design since if it was held for say, the entirety of input processing, it would be highly contested.

In order to prevent lock order inversion, every time `sctp_flags` is read we'd need to unlock the STCB lock before acquiring the INP lock. I don't imagine this is feasible, since I assume in most cases the caller of a given function that reads `sctp_flags` isn't expecting for its STCB lock to be released temporarily before its work with it is done.

So it seems that the atomic approach is only one feasible. Meaning I'd replace every read with:

`if (atomic_load_int(&stcb->sctp_ep->sctp_flags, 0) & SCTP_PCB_FLAGS_TO_READ)`

And every write with:

```
do {
  flags = atomic_load_int(&stcb->sctp_ep->sctp_flags, 0);
} while (!atomic_cmpset_int(&stcb->sctp_ep->sctp_flags,
                            flags, (flags | SCTP_PCB_FLAGS_TO_WRITE)));
```

Does this sound reasonable @tuexen?

**Comment by taylor-b** [comment]
@tuexen Ping; just want to make sure this is an acceptable approach before submitting a PR.
